monai/transforms/utility/array.py

```diff
@@ -1715,3 +1715,101 @@ def __call__(self, img: NdarrayOrTensor, meta_dict: Mapping | None = None) -> Nd
         if self._do_transform:
             img = self.filter(img)
         return img
+
+
+class CoordinateTransform(InvertibleTransform, Transform):
+    """
+    Transform points between image coordinates and world coordinates.
+
+    Args:
+        dtype: The desired data type for the output.
+        mode: Specifies the direction of transformation. This should be an instance of
+            :py:class:`CoordinateTransformMode` enum, with either 'IMAGE_TO_WORLD' or 'WORLD_TO_IMAGE' as the value.
+        affine_lps_to_ras: Defaults to ``False``. Set this to ``True`` if: 1) The image is read by ITKReader,
+            and 2) The ITKReader has `affine_lps_to_ras=True`, and 3) The data is in world coordinates.
+            This ensures that the affine transformation between LPS (left-posterior-superior) and RAS
+            (right-anterior-superior) coordinate systems is correctly applied.
+    """
+
+    def __init__(
+        self,
+        dtype: DtypeLike = torch.float64,
+        mode=CoordinateTransformMode.IMAGE_TO_WORLD,
+        affine_lps_to_ras: bool = False,
+    ) -> None:
+        self.dtype = dtype
+        self.mode = mode
+        self.affine_lps_to_ras = affine_lps_to_ras
+
+
+    @staticmethod
+    def apply_affine_to_points(data, affine, dtype):
```

Comment:
I think apply_affine_to_points should be a standalone utility implementation (see PR #7955) as we will need this functionality deep in the pipeline execution.
I also think that we would be better off loading points into a form suitable for multiplication with transformation matrices (i.e. (x,y,1) or (x,y,z,1) rather than constantly adding and removing the final element (see monai/transforms/lazy/functional.py L299 in PR #7955)

Reply:
> apply_affine_to_points should be a standalone utility implementation

`apply_affine_to_points ` can be reused as it's a staticmethod. I also reused it in the Flip transform.

> I also think that we would be better off loading points into a form suitable for multiplication with transformation matrices (i.e. (x,y,1) or (x,y,z,1) rather than constantly adding and removing the final element

I didn't get the point for this one. The axis is only added and removed in this transform, not affect any place else.

Reply:
Ahh, but the points themselves will be transformed on multiple occasions in a typical pipeline. This means that each time, they are transformed, the final element needs to be added and then removed. If we just add it up front and only remove it on export, then there is less tensor churn

Reply:
Yes, Ben. I know your points. But after deeper investigation, I feel we didn't enhance other spatial transforms to be compatible with the geometric data, we can just use this transform once to convert the point back to the transformed image space, right?

Reply:
I don't think so.
Example pipeline:
LoadImaged, LoadGeometryd, Rotated, Zoomd, Spacingd
Rotated will need to transform the points, Zoomd will need to transform the points, Zoomd will need to transform the points, and Spacingd will need to transform the raster data but apply a "no-op" to the geometry data so the applied entries are all the same length. We probably don't want to modify the geometry tensors twice by adding a tensor of [1s] then removing it after each operation. Please look at PR #7955. I've enabled geometry for many of the spatial transforms already.

Reply:
Closing this as we had already discussed this in meetings. Please feel free to re-open if there are still questions.